File: var/www/html/api_semantic.py
#!/usr/bin/env python3
################################################################
#  Copyright (C) 2016-2026 Fawcett Innovations LLC             #
#                                                              #
#  SPDX-License-Identifier: GPL-2.0-only                       #
#                                                              #
#  This program is free software; you can redistribute it      #
#  and/or modify it under the terms of the GNU General Public  #
#  License as published by the Free Software Foundation;       #
#  version 2 of the License, and no other version.             #
#                                                              #
#  This program is distributed in the hope that it will be     #
#  useful, but WITHOUT ANY WARRANTY; without even the implied  #
#  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR     #
#  PURPOSE.  See the GNU General Public License for details.   #
#                                                              #
#  See COPYRIGHT and LICENSE at the root of this tree.         #
################################################################
import os
import json
import time
import urllib.request
import urllib.parse

API_BASE = os.environ.get("FROGNET_API_BASE", "http://databasehost.frognet/api.php")
TIMEOUT = float(os.environ.get("FROGNET_API_TIMEOUT", "2.5"))

def fetch_json(url: str):
    req = urllib.request.Request(url, headers={"Accept":"application/json"})
    with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
        return json.loads(r.read().decode("utf-8", "replace"))

def latest_sensor_data(sensor_type: str, metric_name: str, limit: int = 50):
    # This assumes your api.php supports:
    #   entity=sensor_data&action=list&SensorType=...&MetricName=...&limit=...
    # If your parameters differ, adjust here only.
    q = {
        "entity": "sensor_data",
        "action": "list",
        "SensorType": sensor_type,
        "MetricName": metric_name,
        "limit": str(int(limit)),
        "order": "desc",
    }
    url = API_BASE + "?" + urllib.parse.urlencode(q)
    data = fetch_json(url)
    # Expected: {"ok":true, "rows":[...]} or raw array
    if isinstance(data, dict) and "rows" in data:
        return data["rows"]
    if isinstance(data, list):
        return data
    return []

def app(environ, start_response):
    path = environ.get("PATH_INFO","/") or "/"
    try:
        if path in ("/", "/summary"):
            proxy_rows = latest_sensor_data("SemanticProxy", "Bytes", limit=200)
            daemon_rows = latest_sensor_data("SemanticDaemon", "RPC", limit=200)

            body = json.dumps({
                "ok": True,
                "ts": time.time(),
                "proxy": proxy_rows,
                "daemon": daemon_rows,
            }, separators=(",",":")).encode("utf-8")

            start_response("200 OK", [
                ("Content-Type","application/json; charset=utf-8"),
                ("Cache-Control","no-store"),
                ("Content-Length", str(len(body))),
            ])
            return [body]

        start_response("404 Not Found", [("Content-Type","application/json")])
        return [b'{"ok":false,"error":"not found"}']

    except Exception as e:
        body = json.dumps({"ok":False,"error":repr(e)}, separators=(",",":")).encode("utf-8")
        start_response("500 Internal Server Error", [
            ("Content-Type","application/json; charset=utf-8"),
            ("Cache-Control","no-store"),
            ("Content-Length", str(len(body))),
        ])
        return [body]

# WSGI entrypoint
application = app

if __name__ == "__main__":
    # simple CGI fallback
    print("Content-Type: application/json; charset=utf-8")
    print("Cache-Control: no-store")
    print()
    out = {"ok": False, "error": "run under WSGI"}
    print(json.dumps(out))

File: var/www/html/api_semantic.php
<?php
/***************************************************************
 *  Copyright (C) 2016-2026 Fawcett Innovations LLC            *
 *                                                             *
 *  SPDX-License-Identifier: GPL-2.0-only                      *
 *                                                             *
 *  This program is free software; you can redistribute it     *
 *  and/or modify it under the terms of the GNU General Public *
 *  License as published by the Free Software Foundation;      *
 *  version 2 of the License, and no other version.            *
 *                                                             *
 *  This program is distributed in the hope that it will be    *
 *  useful, but WITHOUT ANY WARRANTY; without even the implied *
 *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR    *
 *  PURPOSE.  See the GNU General Public License for details.  *
 *                                                             *
 *  See COPYRIGHT and LICENSE at the root of this tree.        *
 **************************************************************/
// /var/www/html/api_semantic.php
// FrogNet telemetry API shim (NO direct DB access). Uses local api.php only.

header('Content-Type: application/json; charset=utf-8');
header('Cache-Control: no-store');

$API_BASE = getenv('FROGNET_API_BASE');
if (!$API_BASE) {
  // default to local api.php on same host
  $scheme = (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off') ? "https" : "http";
  $host = $_SERVER['HTTP_HOST'] ?? 'localhost';
  $API_BASE = $scheme . "://" . $host . "/api.php";
}

$TIMEOUT = floatval(getenv('FROGNET_API_TIMEOUT') ?: "2.5");

function bad($msg, $detail=null, $code=500) {
  http_response_code($code);
  echo json_encode(["ok"=>false,"error"=>$msg,"detail"=>$detail], JSON_UNESCAPED_SLASHES);
  exit;
}

function http_get_json($url, $timeout) {
  $ctx = stream_context_create([
    "http" => [
      "method" => "GET",
      "timeout" => $timeout,
      "header" => "Accept: application/json\r\n",
    ]
  ]);
  $raw = @file_get_contents($url, false, $ctx);
  if ($raw === false) {
    $err = error_get_last();
    return [null, $err ? $err["message"] : "file_get_contents failed"];
  }
  $j = json_decode($raw, true);
  if ($j === null && json_last_error() !== JSON_ERROR_NONE) {
    return [null, "Invalid JSON from api.php (starts: " . substr(preg_replace('/\s+/', ' ', $raw), 0, 200) . ")"];
  }
  return [$j, null];
}

function api_values($api_base, $timeout, $sensorType, $nameLike, $limit=5000) {
  $q = [
    "entity" => "sensors",
    "action" => "values",
    "SensorType" => $sensorType,
    "SensorName__like" => $nameLike,
    "limit" => strval(intval($limit)),
    "parse" => "1",
  ];
  $url = $api_base . "?" . http_build_query($q);
  [$j, $err] = http_get_json($url, $timeout);
  if ($err) return [[], $url, $err];
  $rows = $j["rows"] ?? [];
  if (!is_array($rows)) $rows = [];
  return [$rows, $url, null];
}

function pick_latest_payload($rows) {
  // rows come from sensors/values parse=1, so parsed JSON should be in `data` (best), else try decode jsonData.
  // Choose payload with now_ts if present, else any with since_ts.
  $best = null;
  foreach ($rows as $r) {
    $p = null;
    if (isset($r["data"]) && is_array($r["data"])) $p = $r["data"];
    else if (isset($r["jsonData"]) && is_string($r["jsonData"])) {
      $tmp = json_decode($r["jsonData"], true);
      if (is_array($tmp)) $p = $tmp;
    }
    if (!$p) continue;
    if (isset($p["now_ts"])) return $p;
    if ($best === null) $best = $p;
  }
  return $best ?: new stdClass();
}

// --- router ---
$path = $_SERVER["PATH_INFO"] ?? "/";
if ($path === "/") $path = "/summary";

if ($path !== "/summary") {
  bad("not found", ["path"=>$path], 404);
}

$proxyLike  = "%\.SemanticProxy\.Bytes";
$daemonLike = "%\.SemanticDaemon\.Bytes";

[$proxyRows,  $proxyUrl,  $proxyErr]  = api_values($API_BASE, $TIMEOUT, "SemanticProxy",  $proxyLike);
[$daemonRows, $daemonUrl, $daemonErr] = api_values($API_BASE, $TIMEOUT, "SemanticDaemon", $daemonLike);

$out = [
  "ok" => true,
  "ts" => time(),
  "api_base" => $API_BASE,
  "sources" => [
    "proxy"  => ["url"=>$proxyUrl,  "error"=>$proxyErr],
    "daemon" => ["url"=>$daemonUrl, "error"=>$daemonErr],
  ],
  "proxy_rows" => $proxyRows,
  "daemon_rows" => $daemonRows,
  "proxy_latest" => pick_latest_payload($proxyRows),
  "daemon_latest" => pick_latest_payload($daemonRows),
];

echo json_encode($out, JSON_UNESCAPED_SLASHES);
